File: gamepkg/persistence.py
import json, os
from .constants import STATS_PATH, CONFIG_PATH

# Multi-profile wrappers
STATS_WRAPPER  = {"profiles": {}}                   # name -> stats dict
CONFIG_WRAPPER = {"profiles": {}, "last_profile": None}  # name -> config dict

def load_stats(path: str = STATS_PATH):
    global STATS_WRAPPER
    if os.path.exists(path):
        try:
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
            if "profiles" not in data:  # migrate legacy single-player schema
                STATS_WRAPPER = {"profiles": {"Default": data}}
            else:
                STATS_WRAPPER = data
        except Exception:
            STATS_WRAPPER = {"profiles": {}}
    else:
        STATS_WRAPPER = {"profiles": {}}

def save_stats(path: str = STATS_PATH):
    try:
        with open(path, "w", encoding="utf-8") as f:
            json.dump(STATS_WRAPPER, f, indent=2)
    except Exception:
        pass  # non-fatal

def load_config(path: str = CONFIG_PATH):
    global CONFIG_WRAPPER
    if os.path.exists(path):
        try:
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
            if "profiles" not in data:  # migrate legacy single-player schema
                CONFIG_WRAPPER = {"profiles": {"Default": data}, "last_profile": "Default"}
            else:
                data.setdefault("profiles", {})
                data.setdefault("last_profile", None)
                CONFIG_WRAPPER = data
        except Exception:
            CONFIG_WRAPPER = {"profiles": {}, "last_profile": None}
    else:
        CONFIG_WRAPPER = {"profiles": {}, "last_profile": None}

def save_config(path: str = CONFIG_PATH):
    try:
        with open(path, "w", encoding="utf-8") as f:
            json.dump(CONFIG_WRAPPER, f, indent=2)
    except Exception:
        pass  # non-fatal

File: web/js/persistence.js
import { defaultStats, defaultConfig } from './constants.js';

const STATS_KEY  = 'rps_stats';
const CONFIG_KEY = 'rps_config';

// Module-level state — mirrors the Python STATS_WRAPPER / CONFIG_WRAPPER globals
export const state = {
  stats:  { profiles: {} },
  config: { profiles: {}, last_profile: null },
};

export function loadAll() {
  try {
    const s = localStorage.getItem(STATS_KEY);
    if (s) {
      const data = JSON.parse(s);
      // migrate legacy single-player schema
      state.stats = ('profiles' in data) ? data : { profiles: { Default: data } };
    }
  } catch (_) { /* non-fatal */ }

  try {
    const c = localStorage.getItem(CONFIG_KEY);
    if (c) {
      const data = JSON.parse(c);
      if ('profiles' in data) {
        state.config = { profiles: {}, last_profile: null, ...data };
      } else {
        state.config = { profiles: { Default: data }, last_profile: 'Default' };
      }
    }
  } catch (_) { /* non-fatal */ }
}

export function saveAll() {
  try { localStorage.setItem(STATS_KEY,  JSON.stringify(state.stats));  } catch (_) {}
  try { localStorage.setItem(CONFIG_KEY, JSON.stringify(state.config)); } catch (_) {}
}

/** Ensure a profile exists in both wrappers; returns { stats, config } for that profile. */
export function ensureProfile(name) {
  if (!state.stats.profiles[name])  state.stats.profiles[name]  = defaultStats();
  if (!state.config.profiles[name]) state.config.profiles[name] = defaultConfig();
  return {
    stats:  state.stats.profiles[name],
    config: state.config.profiles[name],
  };
}

export function profileNames() {
  const names = new Set([
    ...Object.keys(state.stats.profiles),
    ...Object.keys(state.config.profiles),
  ]);
  return [...names].sort();
}
